apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/concurrent/SequenceNumberTracker.java

```diff
@@ -0,0 +1,78 @@
+ * they left off on previous calls. We track these sequence numbers in the database so that our job
+ * knows what sequence number to send.
+ *
+ * <p>Concurrent writes complicate sequence number tracking since there is no guaranteed ordering of
+ * how claims are written to the database. The best we can do is track the highest known sequence
+ * number for which there are no unwritten records with a lower sequence number.
+ *
+ * <p>This thread safe container tracks batches of sequence numbers as they are being written to the
+ * database. At any given time the sequence numbers that are currently being written to the database
+ * are maintained in a Set. They are removed from the Set as they are written. At all times the next
+ * unwritten sequence number is available as the smallest sequence number in the Set. If the Set is
+ * empty then there are no batches in progress and instead the highest number that has ever been
+ * written is known to be safe.
+ */
+@ThreadSafe
+public class SequenceNumberTracker {
+  // Use a TreeSet so that values are sorted in ascending order.
+  private final Set<Long> activeSequenceNumbers = new TreeSet<>();
+  private long maxSequenceNumber;
+
+  /**
+   * Creates an object using the specified startingSequenceNumber. This number will be returned
+   * until any batches have been started.
+   *
+   * @param startingSequenceNumber default value for getNextSequenceNumber()
+   */
+  public SequenceNumberTracker(long startingSequenceNumber) {
+    this.maxSequenceNumber = startingSequenceNumber;
+  }
+
+  /**
+   * Adds an active sequence number to the set. This number will need to be removed once the record
+   * has been successfully written.
+   *
+   * @param sequenceNumber of a record that has been queued for writing to the database
+   */
+  public synchronized void addActiveSequenceNumber(long sequenceNumber) {
+    activeSequenceNumbers.add(sequenceNumber);
+    if (sequenceNumber > maxSequenceNumber) {
+      maxSequenceNumber = sequenceNumber;
+    }
+  }
+
+  /**
+   * Removes a sequence number that has been successfully written to the database.
+   *
+   * @param sequenceNumber of a record that has been successfully stored in the database
+   */
+  public synchronized void removeWrittenSequenceNumber(long sequenceNumber) {
+    activeSequenceNumbers.remove(sequenceNumber);
+  }
+
+  /**
+   * Gets the current sequence number for which we know there are no unwritten records with a lower
+   * sequence number.
+   *
+   * @return the best sequence number to send to the RDA API when fetching claims
+   */
+  public synchronized long getHighestWrittenSequenceNumber() {
+    if (activeSequenceNumbers.size() > 0) {
+      // Since the set is ordered the first value is lowest unwritten value.
+      // We return one less so that caller knows to resume at our first unwritten value.
+      return activeSequenceNumbers.iterator().next() - 1;
```

Comment:
I'm pretty sure this doesn't matter, but I think there is a chance here of returning a value that has yet to be recorded in the DB.  There is no guarantee that we will receive a message for every possible sequence number value.  If for example RDA never sends a message for sequence number 100, but at the time of running this function the smallest number in the set for 101, this function will return 100.  Again, this seems fine if all we need is a number to know where to continue from, but if there is some logic that assumes this value DOES exist in our DB we might have a problem.

Reply:
Your analysis is correct.  The purpose of this is to track safe values for the sequence number column in the progress table.  It's not to guarantee that the number is always a value received from the API but it will be a value between the value in the database when the pipeline started and the lowest unwritten value that we have received while the pipeline was running.

Any logic that depends on the value in this column actually existing in a record within the claims tables would be a pretty serious design bug.  There is no integrity constraint relating to this table to records in the claim tables.  Also there is aonly a single column holding a sequence number so no good way to create any such constraint.  No cxonstraint is needed though because this table is really just a scratch pad to allow the pipeline to track its best known resumption point as it runs and between runs.

Even if someone tried to work out a way to have a constraint like that it would be doomed to cause problems simply because the transaction that writes the sequence numbers to the progress table is decoupled from the transactions that update the claims themselves.  It's not unlikely that the sequence number value to be written at this moment might already have been made obsolete by another transaction updating the claim that originally had that sequence number.  There is no harm in that since in the worst case of the pipeline crashing at that moment and restarting it would just resume at the older point and reprocess something it had already done but not recorded in the progress table yet.

That's by design since it safely allows all of the claim update transactions to work within just the claims tables.  They cannot conflict with each other while trying to also update the same record of the progress table in the same transaction.